src/cancelq/cancellable.h

```diff
@@ -0,0 +1,78 @@
+#ifndef CANCELLABLE_H
+#define CANCELLABLE_H
+
+#include <stddef.h>
+#include <stdbool.h>
+
+#define NOT_CANCELLABLE ((cancellable_id_t)-1)
+#define SCHEDULE_FAIL ((cancellable_id_t)-2)
+
+// The event queue that holds all these events
+typedef struct event_queue_s event_queue_t;
+
+// ID used to cancel cancellable_t
+typedef int cancellable_id_t;
+
+// Return a new event queue that will schedule and deliver events with
+// the eq functions below.
+event_queue_t *eq_new(void);
+
+// Frees an empty event queue. Fails if the queue is not empty.
+bool eq_free(event_queue_t *queue);
+
+// Schedule a cancellable event to be delivered at the appointed
+// time. Returns an ID that can be used to cancel the event. The ID is
+// a limited resource that should be released when it's no longer
+// needed.
+cancellable_id_t eq_schedule(event_queue_t *queue, const void *event,
+                             unsigned long long time);
+
+// Post an uncancellable event to the queue, to be delivered at time.
+bool eq_post(event_queue_t *queue, const void *event, unsigned long long time);
+
+// Returns the next scheduled event for the current time.
+void *eq_next_event(event_queue_t *queue, unsigned long long time);
```

Comment:
Unclear - does this only return zero or one event whose time is less than the supplied time?  That implies everything is in absolute time, not time deltas.

Reply:
I should have said for the provided time. Yes, it returns zero or one event.